[Sema] Add `-fvisibility-global-new-delete=` option

bd1976bris opened the pull request:
By default the implicitly declared replaceable global new and delete
operators are given a default visibility attribute. Previous work, see:
https://reviews.llvm.org/D53787, added
`-fvisibility-global-new-delete-hidden` to change this to a hidden
visibility attribute.

This change adds `-fvisibility-global-new-delete=` which controls
whether (or not) to add an implicit visibility attribute to the implicit
declarations for these functions, and what visibility that attribute
will specify. The option takes 4 values: `force-hidden`,
`force-protected`, `force-default` and `source`. Option values
`force-hidden`, `force-protected` and `force-default` assign hidden,
protected, and default visibilities respectively; the use of the term
force in the value names is designed to imply to a user that the semantics
of this option differ significantly from `-fvisibility=`. An option
value of `source` implies that no implicit attribute is added; without
the attribute the replaceable global new and delete operators behave
normally (like other functions) with respect to visibility attributes,
pragmas and options.

The motivation for the `source` value is to facilitate users who intend
to replace these functions either for a single linkage unit or a limited
set of linkage units. `-fvisibility-global-new-delete=source` can be 
applied globally to the compilations in a build where the existing
`-fvisibility-global-new-delete-hidden` cannot, as it conflicts with a
common pattern where these functions are dynamically imported.

The existing `-fvisibility-global-new-delete-hidden` is now a deprecated
spelling of `-fvisibility-global-new-delete=force-hidden`

The command line help for these options is rendered as:

```
  -fvisibility-global-new-delete-hidden
                          Give global C++ operator new and delete declarations hidden visibility

  -fvisibility-global-new-delete=<value>
                          The visibility for global C++ operator new and delete declarations. If
                          'source' is specified the visibility is not adjusted
```

A release note has been added for these changes.

`-fvisibility-global-new-delete=source` will be set by default for PS5.
PS5 users that want the normal toolchain behaviour will be able to
supply `-fvisibility-global-new-delete=force-default`.

bd1976bris commented:
See #74629 for more on visibility support in PlayStation.

bd1976bris commented:
Noting previous discussions w.r.t. -fvisibility-global-new-delete-hidden on https://reviews.llvm.org/D133266. I have added reviewers from that phab review in case they are interested in this.

pogo59 commented:
From a UI perspective, "add a visibility attribute" doesn't seem all that descriptive. What visibility does it add? Therefore, I suggest `-fvisibility-global-new-delete[=<visibility>]` so the user can ask for what they want, and `-fvisibility-global-new-delete-hidden` equals `-fvisibility-global-new-delete=hidden`. The help text can indicate that the default is `default`.

I'm unclear what the `-fno` variant actually turns into. There's always a visibility, ultimately, right? So wouldn't that be equivalent to the `keep` option in #74629 ? Which suggests not having the `-fno` variant here, and allowing `=keep` instead.

(If the universe would implode given a visibility other than `default` or `hidden` for global new/delete, then constrain the argument to the option to be only `default/hidden/keep`.)

I hope this makes sense...

bd1976bris commented:
> From a UI perspective, "add a visibility attribute" doesn't seem all that descriptive. What visibility does it add? Therefore, I suggest `-fvisibility-global-new-delete[=<visibility>]` so the user can ask for what they want, and `-fvisibility-global-new-delete-hidden` equals `-fvisibility-global-new-delete=hidden`. The help text can indicate that the default is `default`.

Thanks Paul. It's a nice suggestion for the UI. I didn't go down this route because the name `-fvisibility-global-new-delete[=<visibility>]` seems to imply (at least to me) that the option should behave like `-fvisibility[=<visibility>]` . However, the latter only applies to definitions and not to declarations whereas the former would apply to both.

> I'm unclear what the `-fno` variant actually turns into. There's always a visibility, ultimately, right? So wouldn't that be equivalent to the `keep` option in #74629 ? Which suggests not having the `-fno` variant here, and allowing `=keep` instead.

There's a difference! The `keep` option in #74629 is applying to the visibility of the IR globals. However, the replaceable new and delete declarations are given a visibility during SEMA (a `VisibilityAttr`) - as if those declarations were parsed with an explicit visibility attribute. By default that `VisibilityAttr` has `default` visibility but `-fvisibility-global-new-delete-hidden` changes it to `hidden` visibility. The proposed `-fno-visibility-global-new-delete` instructs the compiler not to add a `VisibilityAttr` to these special globals. Therefore, these declarations behaviour like any other declarations w.r.t. visibility options/attributes and pragmas.

petrhosek commented:
I don't think it's very clear from `-f[no-]visibility-global-new-delete` what the option does. How about `-f[no-]visibility-attribute-global-new-delete`?

frobtech commented:
The meaning that we want to make clear is that this toggles the special behavior of forcing the visibility of these symbols despite all the other mechanisms that usually control visibility for all other symbols.  So perhaps `-fforced-global-new-delete-visibility` (or even `-fforced-global-new-delete-visibility=default`) describes the status quo ante, and `-fno-forced-global-new-delete-visibility` describes the new opt-in wherein these symbols are not treated differently than others. (The `=...` variant of the positive form would permit forcing to a different visibility than `default`, though I'm not sure there's a real need for that capability since in `-fno-forced-...` state the usual pragmas et al can arrange for that.)

bd1976bris commented:
Thanks for the responses.

> The meaning that we want to make clear is that this toggles the special behavior of forcing the visibility of these symbols despite all the other mechanisms that usually control visibility for all other symbols.

Agreed. I'm in favour of your suggestion of `-f[no-]forced-global-new-delete-visibility`. Including "forced" in the option name makes it clear that something unusual is going on here and hints that this is going to take precedence over other sources of visibility. I did consider that it might be better to rearrange your suggested option name a bit to `-f[no-]visibility-global-new-delete-forced` to group with other visibility options that start with `-fvisibility`, but `-f[no-]forced-global-new-delete-visibility` is much clearer.  @petrhosek I did consider `-f[no-]visibility-attribute-global-new-delete` but that naming is possibly tied a bit too strongly to the current implementation in the complier - so I think I like the `forced` suggestion better?

>  (The `=...` variant of the positive form would permit forcing to a different visibility than `default`, though I'm not sure there's a real need for that capability since in `-fno-forced-...` state the usual pragmas et al can arrange for that.)

Indeed and the use of the usual pragmas etc is preferable, IMO, as users are already familiar with those - as opposed to the effects of an unusual/niche option.

bd1976bris commented:
I have updated the patch to use `-f[no-]forced-global-new-delete-visibility` as suggested.

pogo59 commented:
Actually I kind of prefer all these options to have `-fvisibility-` as a prefix. Even if it doesn't read quite naturally, it strongly implies that the options are related (which is true) and any lexically sorted list of options will naturally group them together. WDYT @frobtech ?

frobtech commented:
> Actually I kind of prefer all these options to have `-fvisibility-` as a prefix. Even if it doesn't read quite naturally, it strongly implies that the options are related (which is true) and any lexically sorted list of options will naturally group them together. WDYT @frobtech ?

I don't have a strong opinion about the exact spelling. I just recommend that it be as unambiguous as possible, and the use of "forced" in the name seems like a good way to help with that.

zygoloid commented:
How about instead using `-fvisibility-global-new-delete=<value>`? We want a mutually-exclusive set of options here, and we know we want more than two of them, and a flag with an argument value is the way we usually express that.

Perhaps we can use values of `force-hidden`, `force-default` to force the visibility to specific things, and something like `unconstrained` or `source` to say "use the normal visibility calculation"? (It's a bit unfortunate that we can't use `default` for that, given that it means something else...) `-fvisibility-global-new-delete-hidden` could then be converted into a deprecated synonym for `-fvisibility-global-new-delete=force-hidden`.

bd1976bris commented:
@zygoloid - yes, the choice of "default" for the name of one of the visibility values in ELF is annoying :) Thanks for the comments all. I feel that what we might need is a bit of a benevolent dictator here. Does anyone object to Richard's `-fvisibility-global-new-delete=<value>` proposal - I know that `-fvisibility-global-new-delete=force-hidden` is in use so deprecating might be painful for someone. Thinking about it more critically, I think I was a bit put off by what might be involved in deprecating `-fvisibility-global-new-delete=force-hidden` and that might have coloured my view as to what to implement for this.

zygoloid commented:
To be clear: I didn't mean to suggest that `-fvisibility-global-new-delete=force-hidden` itself would be deprecated. I think supporting it would have low cost and it may be exactly what is desired in some cases, and the previous spelling [has quite a few uses](https://www.google.com/search?q=%22-fvisibility-global-new-delete-hidden%22). Instead, what I'd suggest is that we add the more-general mechanism of `-fvisibility-global-new-delete=<value>` and make the existing `-fvisibility-global-new-delete-hidden` equivalent to `-fvisibility-global-new-delete=force-hidden`, except that the former spelling is deprecated and the latter spelling is not.

bd1976bris commented:
Given that there have been no objections I have modified this to match @zygoloid's suggestion of `-fvisibility-global-new-delete=<value>`. Please take another look.

pogo59 commented on clang/lib/Driver/ToolChains/PS4CPU.cpp:
```diff
@@ -359,6 +359,12 @@ void toolchains::PS4PS5Base::addClangTargetOptions(
 
   CC1Args.push_back("-fno-use-init-array");
 
+  // Default to -fglobal-new-delete-visibility=source for PS5.
```
Spelling in comment needs to match the option spelling.

bd1976bris replied:
Thanks - now using the correct spelling.

pogo59 reviewed (commented):
I'm okay with the PS5-specific bits. Other people should have a chance to chime in so I won't mark it approved.

MaskRay commented on clang/test/Driver/visibility-global-new-delete.cl:
```diff
@@ -0,0 +1,52 @@
+/// Check driver handling for "-fvisibility-global-new-delete-hidden" and "-fvisibility-global-new-delete=".
+
+/// These options are not added by default.
+// RUN: %clang -### -target x86_64-unknown-unknown -x cl -c -emit-llvm %s 2>&1 | \
```
`--target=` instead of deprecated `-target ` for new tests

MaskRay commented on clang/test/CodeGenCXX/visibility-global-new-delete.cpp:
```diff
@@ -0,0 +1,14 @@
+
+// Definition which inherits visibility from the implicit compiler generated declaration.
+void operator delete(void*) throw() {}
+// CHECK: define [[LINKAGE]] void @_ZdlPv
```
Add a test for `operator new`?

`-fvisibility-global-new-delete-hidden`, from https://reviews.llvm.org/D53787, does not have a test. Thanks for adding the test

bd1976bris replied:
I don't think there is much value in adding operator new as these declarations are all handled in the same way. I think minimal testing here is better.

MaskRay commented on clang/include/clang/Driver/Options.td:
```diff
@@ -3864,9 +3864,16 @@ def fvisibility_ms_compat : Flag<["-"], "fvisibility-ms-compat">, Group<f_Group>
+def fvisibility_global_new_delete_EQ : Joined<["-"], "fvisibility-global-new-delete=">, Group<f_Group>,
   Visibility<[ClangOption, CC1Option]>,
-  MarshallingInfoFlag<LangOpts<"GlobalAllocationFunctionVisibilityHidden">>;
+  HelpText<"The visibility for global C++ operator new and delete declarations definitions. If source is specified the visibility is not adjusted">,
```
Single quotes around `source`

MaskRay reviewed (approved):
Consider adding a note to clang/docs/ReleaseNotes.rst. With a few nits this LGTM.

bd1976bris committed:
[Sema] Add `-fvisibility-global-new-delete=` option

By default the implicitly declared replaceable global new and delete
operators are given a default visibility attribute. Previous work, see:
https://reviews.llvm.org/D53787, added
`-fvisibility-global-new-delete-hidden` to change this to a hidden
visibility attribute.

This change adds `-fvisibility-global-new-delete=` which controls
whether (or not) to add an implicit visibility attribute to the implicit
declarations for these functions, and what visibility that attribute
will specify. The option takes 4 values: `force-hidden`,
`force-protected`, `force-default` and `source`. Option values
`force-hidden`, `force-protected` and `force-default` assign hidden,
protected, and default visibilities respectively; the use of the term
force in the value names is designed to imply to a user that the semantics
of this option differ significantly from `-fvisibility=`. An option
value of `source` implies that no implicit attribute is added; without
the attribute the replaceable global new and delete operators behave
normally (like other functions) with respect to visibility attributes,
pragmas and options.

The motivation for the `source` value is to facilitate users who intend
to replace these functions either for a single linkage unit or a limited
set of linkage units. `-fvisibility-global-new-delete=source` can be
applied globally to the compilations in a build where the existing
`-fvisibility-global-new-delete-hidden` cannot, as it conflicts with a
common pattern where these functions are dynamically imported.

The existing `-fvisibility-global-new-delete-hidden` is now a deprecated
spelling of `-fvisibility-global-new-delete=force-hidden`

A release note has been added for these changes.

`-fvisibility-global-new-delete=source` will be set by default for PS5.
PS5 users that want the normal toolchain behaviour will be able to
supply `-fvisibility-global-new-delete=force-default`.

bd1976bris commented:
I rebased the changes onto the latest main to resolve the merge conflicts - these were just textual not semantic.

MaskRay commented:
> I rebased the changes onto the latest main to resolve the merge conflicts - these were just textual not semantic.

You may like getcord/spr :)
https://maskray.me/blog/2023-09-09-reflections-on-llvm-switch-to-github-pull-requests#my-workflow

bd1976bris commented:
> > I rebased the changes onto the latest main to resolve the merge conflicts - these were just textual not semantic.
> 
> You may like getcord/spr :) https://maskray.me/blog/2023-09-09-reflections-on-llvm-switch-to-github-pull-requests#my-workflow

Thanks for the link to the write up. Very much appreciated. I need to improve my workflow now that everything is via github. I don't have much time for GUI interfaces... you'll notice that I managed to repeat the header line in the commit comment when committing through the GUI: https://github.com/llvm/llvm-project/commit/27ce26b06655cfece3d54b30e442ef93d3e78ac7 🥇
